core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -450,56 +467,53 @@ private long updateVersion(Engine.Operation op, long currentVersion, long expect
-            index(index, indexWriter);
+            index.setCreated(true);
+            try {
+                index(index.docs(), indexWriter);
+            } catch (IOException e) {
+                index.setFailure(new IndexFailedEngineException(shardId, index.type(), index.id(), e));
+            }
         } else {
-            created = update(index, versionValue, indexWriter);
+            if (versionValue != null) {
+                index.setCreated(versionValue.delete()); // we have a delete which is not GC'ed...
+            }
+            try {
+                update(index.uid(), index.docs(), indexWriter);
+            } catch (IOException e) {
+                index.setFailure(new IndexFailedEngineException(shardId, index.type(), index.id(), e));
+            }
         }
-        return created;
     }
 
-    private static void index(final Index index, final IndexWriter indexWriter) throws IOException {
-        if (index.docs().size() > 1) {
-            indexWriter.addDocuments(index.docs());
+    private static void index(final List<ParseContext.Document> docs, final IndexWriter indexWriter) throws IOException {
+        if (docs.size() > 1) {
+            indexWriter.addDocuments(docs);
         } else {
-            indexWriter.addDocument(index.docs().get(0));
+            indexWriter.addDocument(docs.get(0));
         }
     }
 
-    private static boolean update(final Index index, final VersionValue versionValue, final IndexWriter indexWriter) throws IOException {
-        final boolean created;
-        if (versionValue != null) {
-            created = versionValue.delete(); // we have a delete which is not GC'ed...
+    private static void update(final Term uid, final List<ParseContext.Document> docs, final IndexWriter indexWriter) throws IOException {
+        if (docs.size() > 1) {
+            indexWriter.updateDocuments(uid, docs);
         } else {
-            created = false;
+            indexWriter.updateDocument(uid, docs.get(0));
         }
-        if (index.docs().size() > 1) {
-            indexWriter.updateDocuments(index.uid(), index.docs());
-        } else {
-            indexWriter.updateDocument(index.uid(), index.docs().get(0));
-        }
-        return created;
     }
 
     @Override
-    public void delete(Delete delete) throws EngineException {
+    public void delete(Delete delete) {
         try (ReleasableLock lock = readLock.acquire()) {
             ensureOpen();
             // NOTE: we don't throttle this when merges fall behind because delete-by-id does not create new segments:
             innerDelete(delete);
-        } catch (IllegalStateException | IOException e) {
-            try {
-                maybeFailEngine("delete", e);
-            } catch (Exception inner) {
-                e.addSuppressed(inner);
-            }
-            throw new DeleteFailedEngineException(shardId, delete, e);
+        } catch (OutOfMemoryError | IllegalStateException t) {
```

Comment:
Same comment here. This will also swallow an `OutOfMemoryError` which we definitely do not want to do and it's questionable to me that we catch it all?